File: 프로그래머스/lv3_네트워크.cpp
#include <string>
#include <vector>
#include <iostream>

using namespace std;

bool visit[200];
int cnt;

void DFS(vector<vector<int>>& computers, int n){

    if( visit[n] == false ){
        visit[n] = true;
        
        // 연결된 네트워크 탐색
        for ( int i=0; i<computers[n].size(); i++){
            if( computers[n][i] == 1 && i != n ){
                DFS(computers, i);
            }
        }
    }

}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    for(int i=0; i<n; i++){
        if( visit[i] == false ){
            cnt++; // 새로운 네트워크를 탐색할때마다, cnt 상승
            DFS(computers, i);
        }
    }

    answer = cnt;
    
    return answer;
}

int main(void){
    int n = 3;
    vector<vector<int>> computers = {{1,1,0}, {1,1,0}, {0,0,1}};
    int ret;

    ret = solution(n, computers);

    cout << ret;

    return 0;
}

File: 프로그래머스/lv3_네트워크.py
def DFS(i, computers):
    if visit[i] is False:
        visit[i] = True
    else:
        return
    
    for j in range(0, len(computers[i])):
        if computers[i][j] == 1:
            DFS(j, computers)

def solution(n, computers):
    global answer
    answer = 0
    global visit
    visit = [ False for i in range(0,len(computers)) ]
    
    for i in range(0, len(computers)):
        if( visit [i] is False ):
            DFS(i, computers)
            answer += 1
    
    return answer

if __name__ == "__main__":
    ret = solution(3, [[1, 1, 0], [1, 1, 0], [0, 0, 1]])
    print(ret)